File: modules/google-earth-engine/docker/sepal-ee/sepal/ee/rx/export.py
from typing import Union

import ee
from rx import concat
from rx.operators import flat_map
from sepal.rx.workqueue import WorkQueue
from sepal.task.rx.observables import progress

from .asset import delete_asset
from .observables import enqueue, execute
from .task import execute_task


def export_image_to_asset(
        credentials,
        image: ee.Image,
        description: str = None,
        asset_id: str = None,
        pyramiding_policy: dict = None,
        dimensions=None,
        region: ee.Geometry = None,
        scale: int = None,
        crs: str = None,
        crs_transform: str = None,
        max_pixels: Union[int, float] = None,
        retries: int = 0
):
    asset_id, description = _init_asset_id_and_description(asset_id, description)

    def create_task():
        return ee.batch.Export.image.toAsset(
            image=image,
            description=description,
            assetId=asset_id,
            pyramidingPolicy=pyramiding_policy,
            dimensions=dimensions,
            region=region,
            scale=scale,
            crs=crs,
            crsTransform=crs_transform,
            maxPixels=max_pixels
        )

    return _export_to_asset(
        credentials,
        create_task=create_task,
        description='export_image_to_asset(asset_id={}, description={}'.format(asset_id, description),
        retries=retries
    )


def export_table_to_asset(
        credentials,
        collection: ee.FeatureCollection,
        description: str = None,
        asset_id: str = None,
        retries: int = 0
):
    asset_id, description = _init_asset_id_and_description(asset_id, description)

    def create_task():
        return ee.batch.Export.table.toAsset(
            collection=collection,
            description=description,
            assetId=asset_id
        )

    return _export_to_asset(
        credentials,
        create_task=create_task,
        description='export_table_to_asset(asset_id={}, description={}'.format(asset_id, description),
        retries=retries
    )


def export_image_to_drive(
        credentials,
        image: ee.Image,
        description: str,
        folder: str = None,
        file_name_prefix: str = None,
        dimensions=None,
        region: ee.Geometry = None,
        scale: int = None,
        crs: str = None,
        crs_transform: str = None,
        max_pixels: int = None,
        shard_size: int = None,
        file_dimensions=None,
        skip_empty_tiles=None,
        file_format: str = None,
        format_options: str = None,
        retries: int = 0
):
    def create_task():
        return ee.batch.Export.image.toDrive(
            image=image,
            description=description,
            folder=folder,
            fileNamePrefix=file_name_prefix,
            dimensions=dimensions,
            region=region,
            scale=scale,
            crs=crs,
            crsTransform=crs_transform,
            maxPixels=max_pixels,
            shardSize=shard_size,
            fileDimensions=file_dimensions,
            skipEmptyTiles=skip_empty_tiles,
            fileFormat=file_format,
            formatOptions=format_options
        )

    return _export_to_drive(
        credentials,
        create_task=create_task,
        description='export_image_to_drive(description={}, folder={}, fileNamePrefix={}'.format(
            description, folder, file_name_prefix
        ),
        retries=retries
    )


def export_table_to_drive(
        credentials,
        collection: ee.FeatureCollection,
        description: str = None,
        folder: str = None,
        file_name_prefix: str = None,
        file_format: str = None,
        selectors=None,
        retries: int = 0
):
    def create_task():
        return ee.batch.Export.table.toDrive(
            collection=collection,
            description=description,
            folder=folder,
            fileNamePrefix=file_name_prefix,
            fileFormat=file_format,
            selectors=selectors
        )

    return _export_to_drive(
        credentials,
        create_task=create_task,
        description='export_table_to_drive(description={}, folder={}, fileNamePrefix={}'.format(
            description, folder, file_name_prefix
        ),
        retries=retries
    )


def _init_asset_id_and_description(asset_id, description):
    if asset_id and not description:
        i = asset_id.rfind('/')
        description = asset_id[i + 1:]
    elif description and not asset_id:
        asset_id = '{}/{}'.format(_first_asset_root(), description)
    elif not asset_id and not description:
        raise ValueError('asset_id or description must be specified when exporting asset')
    return asset_id, description


def _first_asset_root():
    asset_roots = ee.data.getAssetRoots()
    if not asset_roots:
        raise Exception('User has no GEE asset roots: {}'.format(ee.Credentials()))
    return asset_roots[0]['id']


def _export_to_asset(credentials, create_task, description, retries):
    def create_observable():
        return execute(credentials, create_task, description=description).pipe(
            flat_map(lambda task: delete_asset(credentials, task.config['assetId']).pipe(
                flat_map(lambda _: execute_task(credentials, task))
            ))
        )

    return _export(
        credentials,
        create_observable=create_observable,
        description=description,
        retries=retries
    )


def _export_to_drive(credentials, create_task, description, retries):
    return _export(
        credentials,
        create_observable=lambda: execute_task(credentials, create_task()),
        description=description,
        retries=retries
    )


# Work (i.e. exports) is grouped by credentials, limiting concurrent exports per credentials
_ee_exports = WorkQueue(
    concurrency_per_group=5,
    description='earth-engine-exports'
)


def _export(credentials, create_observable, description, retries):
    return concat(
        progress(
            default_message='Submitting export task to Google Earth Engine...',
            message_key='tasks.ee.export.pending'
        ),
        enqueue(
            credentials,
            queue=_ee_exports,
            action=create_observable,
            description=description,
            retries=retries
        )
    )

File: modules/task/docker/src/ee/export.js
const ee = require('ee')
const {concat, defer, of} = require('rxjs')
const {first, map, switchMap} = require('rxjs/operators')
const {swallow} = require('sepal/rxjs/operators')

const Path = require('path')
const {limiter$} = require('./eeExportLimiter')
const {Limiter} = require('sepal/service/limiter')
const drive = require('root/drive')
const {initUserBucket$} = require('root/cloudStorage')
const {downloadFromCloudStorage$} = require('root/cloudStorageDownload')
const log = require('sepal/log').getLogger('ee')
const {credentials$} = require('root/credentials')

const runTask$ = require('root/jobs/ee/task/runTask')
const assetRoots$ = require('root/jobs/ee/asset/getAssetRoots')
const deleteAsset$ = require('root/jobs/ee/asset/deleteAsset')

const CONCURRENT_FILE_DOWNLOAD = 3

const drivePath = folder =>
    `SEPAL/exports/${folder}`

const createDriveFolder$ = folder =>
    defer(() => serialize$(
        drive.getFolderByPath$({path: drivePath(folder), create: true})
    )).pipe(
        swallow()
    )

const exportImageToAsset$ = ({
    imageDef,
    description,
    assetId,
    pyramidingPolicy,
    dimensions,
    region,
    scale,
    crs,
    crsTransform,
    maxPixels = 1e13,
    retries = 0
}) => {
    const assetDestination$ = (description, assetId) => {
        if (!assetId && !description)
            throw new Error('description or assetId must be specified')
        description = description || Path.dirname(assetId)
        return assetId
            ? of({description, assetId})
            : assetRoots$({credentials$}).pipe(
            // : assetRoots$().pipe(
                map(assetRoots => {
                    if (!assetRoots || !assetRoots.length)
                        throw new Error('EE account has no asset roots')
                    return ({description, assetId: Path.join(assetRoots[0], description)})
                })
            )
    }

    const exportToAsset$ = ({taskDef, description, assetId, retries}) => {
        if (ee.sepal.getAuthType() === 'SERVICE_ACCOUNT')
            throw new Error('Cannot export to asset using service account.')
        return limiter$(
            concat(
                deleteAsset$({credentials$, assetId}).pipe(swallow()),
                runTask$({credentials$, taskDef, description})
            )
        )
    }

    return assetDestination$(description, assetId).pipe(
        switchMap(({description, assetId}) =>
            exportToAsset$({
                taskDef: {
                    imageDef,
                    method: 'toAsset',
                    args: [
                        // image,
                        description, assetId, pyramidingPolicy, dimensions, region, scale, crs, crsTransform, maxPixels
                    ]
                },
                description: `exportImageToAsset(assetId: ${assetId}, description: ${description})`,
                assetId,
                retries
            })
        )
    )
}

const exportImageToSepal$ = ({
    imageDef,
    folder,
    description,
    downloadDir,
    dimensions,
    region,
    scale,
    crs,
    crsTransform,
    maxPixels = 1e13,
    shardSize,
    fileDimensions,
    skipEmptyTiles,
    fileFormat,
    formatOptions,
    retries
}) => {
    const prefix = description

    const throughCloudStorage$ = () => {
        const exportToCloudStorage$ = ({taskDef, description, retries}) => {
            log.debug('Earth Engine <to Cloud Storage>:', description)
            return limiter$(
                runTask$({credentials$, taskDef, description})
            )
        }

        return initUserBucket$().pipe(
            switchMap(bucketPath => {
                return concat(
                    exportToCloudStorage$({
                        taskDef: {
                            imageDef,
                            method: 'toCloudStorage',
                            args: [
                                // image,
                                description, bucketPath, `${folder}/${prefix}`, dimensions, region, scale, crs,
                                crsTransform, maxPixels, fileDimensions, skipEmptyTiles, fileFormat, formatOptions
                            ]
                        },
                        description: `export to Sepal through CS (${description})`,
                        retries
                    }),
                    downloadFromCloudStorage$({
                        bucketPath,
                        prefix: `${folder}/`,
                        downloadDir,
                        deleteAfterDownload: true
                    })
                )
            })
        )
    }

    const throughDrive$ = () => {
        const exportToDrive$ = ({taskDef, description, folder, retries}) => {
            log.debug('Earth Engine <to Google Drive>:', description)
            return limiter$(
                concat(
                    createDriveFolder$(folder),
                    runTask$({credentials$, taskDef, description})
                )
            )
        }

        const downloadFromDrive$ = ({path, downloadDir}) =>
            drive.downloadSingleFolderByPath$(path, downloadDir, {
                concurrency: CONCURRENT_FILE_DOWNLOAD,
                deleteAfterDownload: true
            })

        return concat(
            exportToDrive$({
                taskDef: {
                    imageDef,
                    method: 'toDrive',
                    args: [
                        // image,
                        description, folder, prefix, dimensions, region, scale, crs,
                        crsTransform, maxPixels, shardSize, fileDimensions, skipEmptyTiles, fileFormat, formatOptions
                    ]
                },
                description: `export to Sepal through Drive (${description})`,
                folder,
                retries
            }),
            downloadFromDrive$({
                path: drivePath(folder),
                downloadDir
            })
        )
    }

    return credentials$.pipe(
        first(),
        switchMap(({userCredentials}) => userCredentials
            ? throughDrive$()
            : throughCloudStorage$()
        )
    )
}

const {limiter$: serialize$} = Limiter({
    name: 'Serializer',
    maxConcurrency: 1
})

module.exports = {exportImageToAsset$, exportImageToSepal$}
